src/process/channel.rs

```diff
@@ -83,51 +144,113 @@ impl Channel {
         }
     }
 
-    /// Waits for associated child process to send ready message
-    /// and return the pid of init process which is forked by child process
-    pub fn wait_for_child_ready(&mut self) -> Result<Pid> {
+    /// Waits for associated intermediate process to send ready message
+    /// and return the pid of init process which is forked by intermediate process
+    pub fn wait_for_intermediate_ready(&mut self) -> Result<Pid> {
         let mut buf = [0; 1];
         self.receiver
             .read_exact(&mut buf)
-            .with_context(|| "Failed to receive a message from the child process.")?;
+            .with_context(|| "Failed to receive a message from the intermediate process.")?;
 
         match Message::from(u8::from_be_bytes(buf)) {
-            Message::ChildReady => {
-                log::debug!("received child ready message");
+            Message::IntermediateReady => {
+                log::debug!("received intermediate ready message");
                 // Read the Pid which will be i32 or 4 bytes.
                 let mut buf = [0; 4];
-                self.receiver
-                    .read_exact(&mut buf)
-                    .with_context(|| "Failed to receive a message from the child process.")?;
+                self.receiver.read_exact(&mut buf).with_context(|| {
+                    "Failed to receive a message from the intermediate process."
+                })?;
 
                 Ok(Pid::from_raw(i32::from_be_bytes(buf)))
             }
             msg => bail!(
-                "receive unexpected message {:?} waiting for child ready",
+                "receive unexpected message {:?} waiting for intermediate ready",
                 msg
             ),
         }
     }
 
-    pub fn close_receiver(&self) -> Result<()> {
+    pub fn close(&self) -> Result<()> {
         unistd::close(self.receiver.as_raw_fd())?;
+        Ok(())
+    }
+}
 
+pub fn init_to_intermediate() -> Result<(SenderInitToIntermediate, ReceiverFromInit)> {
+    let (sender, receiver) = new_pipe()?;
+    Ok((
+        SenderInitToIntermediate { sender },
+        ReceiverFromInit { receiver },
+    ))
+}
+
+pub struct SenderInitToIntermediate {
+    sender: Sender,
+}
+
+impl SenderInitToIntermediate {
```

Comment:
Since we differentiate channel here, the init process don't ready send pid to the intermediate process, so the `init_ready` here don't have to send Pid. In the init process, since it is already inside the pid namespace, it doesn't know what is the real pid outside the pid namespace. The intermediate process knows the pid anyway since it is one doing the fork.